lib/pacer.go

```diff
@@ -0,0 +1,170 @@
+package vegeta
+
+import (
+	"fmt"
+	"math"
+	"time"
+)
+
+// A Pacer defines the rate of hits during an Attack by
+// returning the duration an Attacker should wait until hitting
+// the next Target. If the second return value is true, the
+// attack will terminate.
+type Pacer interface {
+	Pace(elapsedTime time.Duration, elapsedHits uint64) (sleep time.Duration, stop bool)
+}
+
+// A PacerFunc is a function adapter type that implements the
+// Pacer interface.
+type PacerFunc func(time.Duration, uint64) (time.Duration, bool)
+
+// Pace implements the Pacer interface.
+func (f PacerFunc) Pace(elapsedTime time.Duration, elapsedHits uint64) (time.Duration, bool) {
+	return f(elapsedTime, elapsedHits)
+}
+
+// Rate sends a constant rate of hits to the target.
+type Rate struct {
+	Freq int           // Frequency (number of occurrences) per ...
+	Per  time.Duration // Time unit, usually 1s
+}
+
+// Pace determines the length of time to sleep until the next hit is sent.
+func (r Rate) Pace(elapsedTime time.Duration, elapsedHits uint64) (time.Duration, bool) {
+	if r.Per <= 0 || r.Freq <= 0 {
+		// If Rate configuration is invalid, stop the attack.
+		return 0, true
+	}
+	interval := uint64(r.Per.Nanoseconds() / int64(r.Freq))
+	delta := time.Duration((elapsedHits + 1) * interval)
```

Comment:
watch out for `uint64` overflow - stumbled on it while testing

Reply:
:O

The docs say time.Duration is good for durations of up to ~290 years. Practically speaking, if you are anywhere near the overflow point, Pace is not going to be doing anything useful. But I can add a check for e.g. `INT_MAX / (elapsedHits + 1) > interval` to detect overflows and return `0, true` in this case if you think it's necessary.

Reply:
> Practically speaking, if you are anywhere near the overflow point, Pace is not going to be doing anything useful

You are right, I was testing the full range of input arguments, and it is more a thing of a taste ...

I think tests should drive such decisions (whether to add check or not), as it would be simpler to write them (provided that you wrote good quality tests :) ).